src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala

```diff
@@ -0,0 +1,218 @@
+package org.broadinstitute.dsde.workbench.leonardo.auth
+
+import akka.http.scaladsl.model.StatusCodes
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
+import com.google.api.client.json.jackson2.JacksonFactory
+import com.google.api.services.plus.PlusScopes
+import com.google.auth.oauth2.ServiceAccountCredentials
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import com.google.gson.internal.LinkedTreeMap
+import com.typesafe.config.Config
+import io.swagger.client.ApiClient
+import io.swagger.client.api.ResourcesApi
+import io.swagger.client.api.GoogleApi
+import java.io.{ByteArrayInputStream, File}
+import java.util.concurrent.TimeUnit
+import com.google.gson.Gson
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.dsde.workbench.leonardo.model._
+import org.broadinstitute.dsde.workbench.leonardo.model.Actions._
+import org.broadinstitute.dsde.workbench.leonardo.model.NotebookClusterActions._
+import org.broadinstitute.dsde.workbench.leonardo.model.ProjectActions._
+import org.broadinstitute.dsde.workbench.model.WorkbenchEmail
+import net.ceedubs.ficus.Ficus._
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+
+case class UserEmailAndProject(userEmail: String, googleProject: String)
+
+case class NotebookActionError(action: Action) extends
+  LeoException(s"${action.toString} was not recognized", StatusCodes.NotFound)
+
+
+class SamAuthProvider(authConfig: Config, serviceAccountProvider: ServiceAccountProvider) extends LeoAuthProvider(authConfig, serviceAccountProvider) with LazyLogging {
```

Comment:
I don't think the superclass needs to know about `ServiceAccountProvider`. Can we leave it only in the subclass, e.g.
```
class SamAuthProvider(authConfig: Config, val serviceAccountProvider: ServiceAccountProvider) extends LeoAuthProvider(authConfig) with LazyLogging
```
so as to make the interface as flexible as possible.

Reply:
if it's in the subclass how does a generic constructor know what parameters to pass?

Reply:
I think that's a good idea!

Reply:
Hm -- good point that it needs to be in the subclass in order for our reflection-based `constructProvider` to work. But then we're leaking data into the subclass that isn't needed, ugh.

We could just leave this as is and call it housekeeping to figure out a better way of doing dependency injection in these providers....

Reply:
I agree it's a bit less encapsulated but "an auth provider might need access to either the user or leo SAs" didn't seem like an entirely unreasonable use case.

Perhaps add a provider registry to housekeeping.